common/common_test.go

```diff
@@ -0,0 +1,187 @@
+	step += 1
+	// Test retry logic (for the moment, it's very simple, just one retry).
+	switch step {
+	case 1:
+		return getblockchaininfoReply, errors.New("first failure")
+	}
+	if sleepCount != 1 || sleepDuration != 15*time.Second {
+		testT.Error("unexpected sleeps", sleepCount, sleepDuration)
+	}
+	return getblockchaininfoReply, nil
+}
+
+func TestGetSaplingInfo(t *testing.T) {
+	testT = t
+	saplingHeight, blockHeight, chainName, branchID :=
+		GetSaplingInfo(getblockchaininfoStub, sleepStub, testlog)
+
+	// Ensure the retry happened as expected
+	logFile, err := ioutil.ReadFile("test-log")
+	if err != nil {
+		t.Fatal("Cannot read test-log", err)
+	}
+	logStr := string(logFile)
+	if !strings.Contains(logStr, "retrying") {
+		t.Fatal("Cannot find retrying in test-log")
+	}
+	if !strings.Contains(logStr, "retry=1") {
+		t.Fatal("Cannot find retry=1 in test-log")
+	}
+
+	// Check the success case (second attempt)
+	if saplingHeight != 419200 {
+		t.Error("unexpected saplingHeight", saplingHeight)
+	}
+	if blockHeight != 677713 {
+		t.Error("unexpected blockHeight", blockHeight)
+	}
+	if chainName != "main" {
+		t.Error("unexpected chainName", chainName)
+	}
+	if branchID != "2bb40e60" {
+		t.Error("unexpected branchID", branchID)
+	}
+
+	if sleepCount != 1 || sleepDuration != 15*time.Second {
+		t.Error("unexpected sleeps", sleepCount, sleepDuration)
+	}
+	step = 0
+	sleepCount = 0
+	sleepDuration = 0
+}
+
+// ------------------------------------------ BlockIngestor()
+
+// There are four test blocks, 0..3
+func getBlockFromRPCStub(method string, params []json.RawMessage) (json.RawMessage, error) {
+	i := step
+	step += 1
+	// Test retry logic (for the moment, it's very simple, just one retry).
+	switch i {
+	case 0:
+		return blocks[0], nil
+	case 1:
+		return blocks[1], nil
+	case 2:
+		// This should cause one sleep (then retry)
+		return nil, errors.New("-8: Block height out of range")
+	case 3:
+		if sleepCount != 1 || sleepDuration != 10*time.Second {
+			testT.Error("unexpected sleeps", sleepCount, sleepDuration)
+		}
+		return blocks[2], nil
+	case 4:
+		// simulate a reorg (this block's prevhash doesn't match previous block's hash)
+		return blocks[1], nil
+	case 5:
+		// it will ask for block 1 (should verify this)
+		return blocks[1], nil
+	case 6:
+		// Should fail to Unmarshal the block, sleep, retry
+		return nil, nil
+	case 7:
+		if sleepCount != 2 || sleepDuration != 20*time.Second {
+			testT.Error("unexpected sleeps", sleepCount, sleepDuration)
+		}
+		return blocks[2], nil
+	}
+	testT.Error("getBlockFromRPCStub called too many times")
+	return nil, nil
+}
+
+func TestBlockIngestor(t *testing.T) {
+	testT = t
+	cache := NewBlockCache(4)
+	BlockIngestor(getBlockFromRPCStub, sleepStub, cache, testlog, 20, 7)
+	if step != 7 {
+		t.Error("unexpected step", step)
+	}
+	step = 0
+}
+
+// ------------------------------------------ BlockIngestor()
```

Comment:
Oops, incorrect comment; I was going to add another test here, will do so in a later commit.